### PR TITLE
Add C# examples to PropertyTweener docs

### LINKED ISSUE #9924: Missing C# examples for PropertyTweener class
**Your Godot version:** 4.3

**Issue description:**
Was looking for more guidance on how to use `SetCustomInterpolator` in C# as a direct lambda expression can't be used directly inline (`.SetCustomInterpolator(t => Math.Pow(t, 0.75f))`).

Took a bit to get to `.SetCustomInterpolator(Callable.From<float, float>(t => Mathf.Pow(t, 0.75f)))`

Would be nice for the page to have the boxes which switch between gdscript and C# like other examples. Thanks!

**URL to the documentation page (if already existing):** https://docs.godotengine.org/en/stable/classes/class_propertytweener.html#class-propertytweener-method-set-custom-interpolator

### PR DESCRIPTION
**Pull request by Bossdell113**
This is something that I am doing for a school project.  I saw an issue posted to the Godot-docs repo about missing C# examples in the PropertyTweener class reference page.  It is labeled as issue #9924 on the docs repo, not the main one, and can be found easily in the good first issues list.

I have added C# examples for the following methods.

as_relative
from
from_current
set_custom_interpolator

I used tools like the code translator as well as cross referencing with other areas of the documentation, primarily the Tween class reference was really helpful, but there are still some parts of this that I am not completely sure about.  I doubt that I am getting this perfect on my first try here, this is the first time I have done anything like this, so feedback for my code and tips for how I can improve my examples and make them more accurate would be greatly appreciated.  I hope I can take any feedback/criticism and use it to make this something that is up to standard.

_Bugsquad edit: closes https://github.com/godotengine/godot-docs/issues/9924_

### CODE REVIEW

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [inline doc/classes/PropertyTweener.xml:26]:
```suggestion
				tween.TweenProperty(this, "position", Vector2.Right * 100.0f, 1.0f).AsRelative();
```
`tween` is the Tween you created, but `Tween` is the class itself.

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [inline doc/classes/PropertyTweener.xml:44]:
```suggestion
				tween.TweenProperty(this, "position", new Vector2(200.0f, 100.0f), 1.0f).From(new Vector2(100.0f, 100.0f));
```

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [inline doc/classes/PropertyTweener.xml:60]:
```suggestion
				tween.TweenProperty(this, "position", new Vector2(200.0f, 100.0f), 1.0f).From(Position);
				tween.TweenProperty(this, "position", new Vector2(200.0f, 100.0f), 1.0f).FromCurrent();
```

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [inline doc/classes/PropertyTweener.xml:96]:
```suggestion
				public TweenCurve(float value)
				{
				    return Curve.SampleBaked(value);
				}
```

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [inline doc/classes/PropertyTweener.xml:90]:
```suggestion
				    Callable tweenCurveCallable = Callable.From((float value) => TweenCurve(value));
				    tween.TweenProperty(this, "position:x", 300.0f, 1.0f).AsRelative().SetCustomInterpolator(tweenCurveCallable);
```
This one compiled and ran for me, though I'm not confident this is the preferred way to create a callable in this case. Needs a review from a C# expert.

(Used this page: [C# Callable](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_differences.html#callable))

**Review by tetrapod00 — DISMISSED** [review]

File: doc/classes/PropertyTweener.xml
Comment (Bossdell113) [reply doc/classes/PropertyTweener.xml:26]: So if I click commit suggestion, it will add in the change that you have suggested?

Comment (tetrapod00) [reply doc/classes/PropertyTweener.xml:26]: Yes, but don't do that just yet. 

Godot likes PRs to have only a single commit, to keep the Git history cleaner. There are two ways to do that:
- Amend your PR as you work on it, described [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request).
- Make fixup commits (including committing suggestions using the GitHub UI), then at the end "squash" all your commits together, described [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase).

You can use either, or both. Usually I find it easier to use amend, and I manually copy and paste suggestions from GitHub to my text editor. But you can also use the GitHub UI to "commit suggestion" and then squash at the end after all the changes are made. It's up to you.

Comment (Bossdell113) [reply doc/classes/PropertyTweener.xml:26]: Ok, then should I wait till the other review that is pending finishes?

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [inline doc/classes/PropertyTweener.xml:19]:
```suggestion
				[codeblocks]
```
I didn't notice originally, but the CI is failing because all these need to be `[codeblocks]` instead of `[codeblock]` when you have multiple languages. Don't forget to change the closing tags to `[/codeblocks]` too.

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [reply doc/classes/PropertyTweener.xml:26]: Up to you. I personally would; usually I like to apply the obvious changes (in this case, everything but the callable one) as soon as possible. The C#-specific review probably won't happen immediately.

Either amending or committing more commits is fine here. In both cases reviewers can see the current state of the PR, with all the previous changes combined.

Comment (tetrapod00) [reply doc/classes/PropertyTweener.xml:26]: Yeah, commiting from the UI is the simplest.

And yeah, for the codeblocks you'll need to make the rest of the changes manually.

Comment (Bossdell113) [reply doc/classes/PropertyTweener.xml:26]: Ok.  I am wondering then if I should just do all manual and keep amending the PR.  Either way, its late where I am so I'll get back to this tomorrow.

**Comment by Bossdell113** [comment]
I have amended this PR with the changes that were suggested.  I don't know if I was supposed to create a new commit in the process but there is a new one here with the changes to the code.

**Comment by tetrapod00** [comment]
Yeah, I'm not sure exactly what happened but it looks like you have two commits instead of a single amended commit. It's not a big deal, just means that you'll have to squash after all the reviews. 

I'm marking all the changes that I suggested and that you applied as "resolved" (except for the callable one). In the future, if someone suggests a change and you apply it exactly as-is, feel free to mark it resolved yourself (but leave it unresolved if you're unsure, or if you didn't apply it exactly as suggested).

Now we're just waiting on a C#-specific review; and once that's done you can rebase/squash into one commit. Good work!

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [inline doc/classes/PropertyTweener.xml:91]:
```suggestion
				    tween.TweenProperty(this, "position:x", 300.0f, 1.0f).AsRelative().SetCustomInterpolator(tweenCurveCallable);
```
The CI is complaining about (at least) this line. It needs to be four spaces, not a tab, for indentation within code examples. Note that the indentation for the XML itself is tabs.

Comment (Bossdell113) [reply doc/classes/PropertyTweener.xml:91]: Oh shoot, are there any other instances of this?

Comment (tetrapod00) [reply doc/classes/PropertyTweener.xml:91]: I don't see any, but you should double check 🙂 

If you plan on making a habit of contributing, there is a way to run the pre-commit locally if you set it up, which should catch most cases like this. See https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#using-clang-format-locally

Comment (tetrapod00) [reply doc/classes/PropertyTweener.xml:91]: Worth noting too that some text editors like VS Code let you distinguish between tabs and spaces. In VS Code it's done when you have text selected. Unfortunately the GitHub UI doesn't do that by default.
Code_WiKm1UeOv3

Comment (Bossdell113) [reply doc/classes/PropertyTweener.xml:91]: I edit in VScode and I'm backspacing and doing four spaces on each line of code where I believe this would be an issue.  The line mentioned here is the only one where it is saying that it has been changed from how it was before.

I believe this is the only one that was tabbed rather than four spaced, and I'm considering just clicking the commit suggestion button because this is so small.

**Review by tetrapod00 — APPROVED** [review]
The C# code examples look good to me, and I tested them in-editor. But this still needs a review from someone more familiar with C# callables and with C#-specific code style for examples.

**Comment by Bossdell113** [comment]
Should we be concerned about the test failure that just occurred?

**Comment by tetrapod00** [comment]
Yep 😄 

I should have waited for *all* the tests to pass, not just a couple of the builds. That one is because the `>` character needs to be escaped within the XML file. You should be able to apply the suggestion it gives you as-is.

**Comment by Bossdell113** [comment]
Could changing that cause any other tests to fail?

**Comment by tetrapod00** [comment]
> Could changing that cause any other tests to fail?

I don't think so, escaping `>` with `&gt;` is standard within the class reference.

I'm not 100% sure why only the one build fails, I think we have them each set to do different things to not waste server time.

**Comment by Bossdell113** [comment]
I made the change 👍

**Comment by Bossdell113** [comment]
And it still failed.  It looks like its freaking out over a single empty line.

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [inline doc/classes/PropertyTweener.xml:100]:
```suggestion
```
This time it's just complaining about an extra empty line.

Comment (Bossdell113) [reply doc/classes/PropertyTweener.xml:100]: I'm just gonna click the button again

File: doc/classes/PropertyTweener.xml
Comment (raulsntos) [inline doc/classes/PropertyTweener.xml:90]: I think this is the syntax we usually prefer.

```suggestion
				    Callable tweenCurveCallable = Callable.From&lt;float, float&gt;(TweenCurve);
```

File: doc/classes/PropertyTweener.xml
Comment (raulsntos) [inline doc/classes/PropertyTweener.xml:94]: Missing return type in this method declaration, and could probably be private.

```suggestion
				private float TweenCurve(float value)
```

File: doc/classes/PropertyTweener.xml
Comment (raulsntos) [inline doc/classes/PropertyTweener.xml:84]:
```suggestion
				public Curve Curve { get; set; }
```

**Review by raulsntos — COMMENTED** [review]
Thanks for contributing to the C# documentation.

File: doc/classes/PropertyTweener.xml
Comment (tetrapod00) [reply doc/classes/PropertyTweener.xml:94]: For clarity, since I said I tested the code: I made a local copy of the code examples in this PR, changed things until they worked, and transcribed the changes into suggestions in this PR. I missed the return value, but it's present in the local version I tested.

**Comment by Bossdell113** [comment]
Ok, i'll add these changes 👍

File: doc/classes/PropertyTweener.xml
Comment (Bossdell113) [reply doc/classes/PropertyTweener.xml]: I have made changes that add these suggestions 👍

**Comment by Bossdell113** [comment]
So all the checks passed, what next?

**Comment by tetrapod00** [comment]
> So all the checks passed, what next?

"All checks have passed" in the UI just means that the automated CI checks have passed. It's a prerequisite for approving a PR but not the only thing that needs to be done. You're still waiting for @raulsntos to come back and approve your changes. Once that happens, you'll have an approval from a docs reviewer (me) and one from a C# reviewer (raulsntos), and with two approvals from the relevant parties, the PR will be fully approved. Some time after that, but not immediately, it will be merged.

**Review by raulsntos — APPROVED** [review]
Looks good to me, thanks!

**Comment by Bossdell113** [comment]
Should I squish the commits then?

**Comment by tetrapod00** [comment]
Yes, please squash! Forgot to mention that in the summary of next steps

**Comment by Bossdell113** [comment]
Ok I think I did it right

**Comment by tetrapod00** [comment]
Looks like you did. I'm not sure if it's a hard requirement for merging, but it would also be nice if you amended the *commit message* to be more descriptive, like the PR title currently is (I changed the PR title for you earlier).

**Comment by Bossdell113** [comment]
So I think I amended it correctly but I don't see the changed description.

**Comment by Bossdell113** [comment]
Wait nevermind I see it

**Comment by tetrapod00** [comment]
image

It's showing up like this for me. Ideally it would be something more like:
```
Add C# examples to PropertyTweener Docs

Docs for the following methods:

as_relative
from
from_current
set_custom_interpolator
```
instead of the current:
```
 Update PropertyTweener.xml

Adding C# examples to PropertyTweener Docs for the following methods

as_relative
from
from_current
set_custom_interpolator
```

If you can't figure this out, it's probably fine, since the PR title is already good and the commit title and description are already reasonably descriptive 🤷. If it's actually a problem then the production team will let you know before merging.

**Comment by Bossdell113** [comment]
Are the other checks supposed to happen again?

**Comment by tetrapod00** [comment]
The checks aren't automatically run, since you're a first-time contributor. I thought the last change you made didn't change the content, so they should still pass. But it doesn't hurt to run the CI again.

**Comment by akien-mga** [comment]
Force pushed to edit the commit message to be clearer.

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Pull request merged**
